File: .graph-engineering/local/scripts/resolve_output_path.py
#!/usr/bin/env python3

import re
import sys
from datetime import datetime, timezone, timedelta

from lib.errors import fail
from lib.paths import get_project_root


def validate_name(value, label):
    """Validate that the value is a non-empty kebab-case identifier."""
    if not value or not re.match(r"^[a-z0-9]+(?:-[a-z0-9]+)*$", value):
        fail(f"{label} MUST be a non-empty kebab-case identifier.")


def get_gmt_minus_five_timestamp():
    """Return current timestamp in GMT-5 formatted as YYYYMMDD-HHMM."""
    tz = timezone(timedelta(hours=-5))
    now = datetime.now(tz)
    return now.strftime("%Y%m%d-%H%M")


def main() -> int:
    if len(sys.argv) != 3:
        fail("Usage: python3 resolve_output_path.py <domain> <job-name>")

    domain = sys.argv[1]
    job_name = sys.argv[2]
    if not domain:
        return 2

    validate_name(domain, "run-name")
    validate_name(job_name, "job-name")

    project_root = get_project_root()
    output_dir = project_root / ".graph-engineering" / "runs" / domain / job_name

    timestamp = get_gmt_minus_five_timestamp()
    output_file = output_dir / f"OUTPUT-{timestamp}.md"

    try:
        output_dir.mkdir(parents=True, exist_ok=True)
    except Exception as e:
        fail(f"Failed to create directory {output_dir}: {e}")

    print(str(output_file))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .graph-engineering/local/scripts/resolve-output-path.mjs
#!/usr/bin/env node

import { mkdir } from "node:fs/promises";
import { join } from "node:path";
import { writeExecutionLog } from "./lib/activity-logs.mjs";
import { resolvePaths } from "./lib/resolve-paths.mjs";

const SCRIPT_NAME = "resolve-output-path";
const paths = resolvePaths(import.meta.url);

const [runName, jobName] = process.argv.slice(2);

function validateName(value, label) {
  const kebabCasePattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

  if (!value || !kebabCasePattern.test(value)) {
    throw new Error(
      `${label} MUST be a non-empty kebab-case identifier.`
    );
  }
}

function getGmtMinusFiveTimestamp() {
  const adjustedTime = new Date(Date.now() - 5 * 60 * 60 * 1000);

  const year = adjustedTime.getUTCFullYear();
  const month = String(adjustedTime.getUTCMonth() + 1).padStart(2, "0");
  const day = String(adjustedTime.getUTCDate()).padStart(2, "0");
  const hour = String(adjustedTime.getUTCHours()).padStart(2, "0");
  const minute = String(adjustedTime.getUTCMinutes()).padStart(2, "0");

  return `${year}${month}${day}-${hour}${minute}`;
}

async function resolveOutputPath() {
  validateName(runName, "run-name");
  validateName(jobName, "job-name");

  const outputDirectory = join(
    paths.projectDataDirectory,
    "runs",
    runName,
    jobName
  );

  const outputFile = join(
    outputDirectory,
    `OUTPUT-${getGmtMinusFiveTimestamp()}.md`
  );

  await mkdir(outputDirectory, { recursive: true });

  await writeExecutionLog({
    scriptName: SCRIPT_NAME,
    logsDirectory: paths.logsDirectory,
    logFile: paths.logFile,
    metadata: {
      run: runName,
      job: jobName,
      output: outputFile,
    },
  });

  console.log(outputFile);
}

try {
  await resolveOutputPath();
} catch (error) {
  console.error(`Failed to resolve output path: ${error.message}`);
  process.exit(1);
}